README.md
# TinyApp Project
TinyApp is a full stack web application built with Node and Express that allows users to shorten long URLs (à la bit.ly).

## Final Product

!["User registration page"](https://github.com/mikel-k-khui/tinyapp/blob/master/docs/register-page.png)
!["URL index page"](https://github.com/mikel-k-khui/tinyapp/blob/master/docs/url_index_page.png)
!["URL edit page"](https://github.com/mikel-k-khui/tinyapp/blob/master/docs/url-edit-page.png)

## Dependencies

- Node.js
- Express
- EJS
- bcrypt
- body-parser
- cookie-session

## Getting Started

- Install all dependencies (using the `npm install` command).
- Run the development web server using the `node express_server.js` command.

HISTORY.md
# *1.1.0 / 2019-08-08 @1805*
Received feedback on major functionality issues due to registration logic not working:
*Root cause and Corrective actions*
  - The if statement to check if an user already exists did not use the correct syntax to test for a null object
    - Code used (getUserByEmail === {})
    - Code should be (Object.keys(getUserByEmail).length as the strict comparison will return the object type instead of value
  - Double-checked all existing codes with {} to ensure all syntax are fixed

Received feedback to move the app.listen() to bottom of the code as best practice:
  - investigating with mentors as to reason
  - current hypothesis is to remove any syntax errors along the way before the server is launched

# *1.0.0 / 2019-08-08 @1415*
Submitted file

test/helpersTest.js
const { assert } = require('chai');

const { getUserByEmail } = require('../helpers.js');

const testUsers = {
  "userRandomID": {
    id: "userRandomID", 
    email: "user@example.com", 
    password: "purple-monkey-dinosaur"
  },
  "user2RandomID": {
    id: "user2RandomID", 
    email: "user2@example.com", 
    password: "dishwasher-funk"
  }
};

describe('getUserByEmail', function() {
  it('should return a user with valid email', function() {
    const user = getUserByEmail("user@example.com", testUsers);
    const expectedOutput = "userRandomID";

    assert.strictEqual(user["email"], testUsers[expectedOutput]["email"]);
  });
  it('should return a user\'s valid password', function() {
    const user = getUserByEmail("user@example.com", testUsers);
    const expectedOutput = "userRandomID";

    assert.strictEqual(user["password"], testUsers[expectedOutput]["password"]);
  });
  it('incorrect id should return {}', function() {
    const user = getUserByEmail("test@example.com", testUsers);
    const expectedOutput = undefined;
  
    assert.strictEqual(user[""], expectedOutput);
  });
});

helpers.js
//Helpers functions
const generateRandomString = function(length) {
  return Math.floor((1 + Math.random()) * 0x1000 * 0x1000).toString(16).substring(1,length);
};

//check if the email eixsts in the users
const getUserByEmail = function(email, usersDatabase) {
  let found = {};
  for (const user in usersDatabase) {
    if (usersDatabase[user].email === email) {
      found = usersDatabase[user];
      break;
    }
  }
  return found;
};

const userIDExist = (id, usersDatabase) => (id in usersDatabase) ? usersDatabase[id].email : undefined;

const regOK = (email, password) => (
  email !== '' &&
  password !== '');

const urlsForUser = function(userID, urlDatabase) {
  let userURLs = {};
  for (const url in urlDatabase) {
    if (urlDatabase[url]["userID"] === userID) {
      userURLs[url] = urlDatabase[url]["longURL"];
    }
  }
  return userURLs;
};

const urlsDetForUser = function(userID, urlDatabase) {
  let userURLs = {};

  for (const url in urlDatabase) {
    if (urlDatabase[url]["userID"] === userID) {
      let urlDet = {};
      urlDet["longURL"] = urlDatabase[url]["longURL"];
      urlDet["created"] = urlDatabase[url]["created"];
      userURLs[url] = urlDet;
    }
  }
  return userURLs;
};

const trackVisitors = function(visDB, url, visCookie) {
  // if (Object.keys(visDB).length === 0) {
  //   //set up an empty object
  //   let urlVisInfo = {};
  //   let visDet = {};
  //   urlVisInfo[url]["totVis"] = 1;
  //   urlVisInfo[url]["uniVis"] = 1;
  //   visDet[visitor_id] = visCookie;
  //   visDet[visCookie][timeStampe] = Date.now();
  //   urlVisInfo[url]["visits"] = visDet;
  // } else {

  // }

  // return visDB;
};

const getVisitorById = function(visDB, visitor_id) {
  // for (const url in visDB) {
  //   let visDet = visDB[url]["visits"];
  //   for (const visitor in visDet) {
  //     if (visitor_id === visitor[])
  //   }

  // }
};

module.exports = {
  getUserByEmail,
  generateRandomString,
  regOK,
  urlsForUser,
  urlsDetForUser,
  userIDExist
};

constant.js
const bcrypt = require('bcrypt');

const users = {
  "1234": {
    id: "1234",
    email: "me@me.com",
    password: bcrypt.hashSync("bob", 10)
  },
  "abcd": {
    id: "abcd",
    email: "i@me.com",
    password: bcrypt.hashSync("bob", 10)
  }
};

const urlDatabase = {
  "b2xVn2": { longURL: "http://www.lighthouselabs.ca", userID: "1234", created: "Thu Aug 8 2019" },
  "9sm5xK": { longURL: "http://www.google.com", userID: "1234", created: "Thu Aug 1 2019"},
  "g82ms9": { longURL: "http://www.triathlon.org", userID: "abcd", created: "Wed Aug 7 2019"}
};

let visitorDatabase = {
  "g82ms9": { totVis: 1, uniVis: 1, visits: { visitor_id: Date.now() - (15 * 60 * 1000) }}

};
// example
// <url key>: { <totVis>: ###, <uniVis>: ###, >visitors>: { <visitor_id>: <random>, <timestamp>: <time>}}

module.exports = { users, urlDatabase, visitorDatabase };

express_server2.js
const express = require("express");
const cookieParser = require("cookie-parser");
const bodyParser = require("body-parser");
const app = express();
const PORT = 8081;

app.set("view engine", "ejs");
//setup Express app to use ejs as templating engine

app.use(bodyParser.urlencoded({extended: true}));
app.use(cookieParser());

const urlDatabase = {
  "b2xVn2": "http://www.lighthouselabs.ca",
  "9sm5xK": "http://www.google.com"
};

const generateRandomString = function() {
  return Math.floor((1 + Math.random()) * 0x1000 * 0x1000).toString(16).substring(1,7);
};

const addURL = function(address) {
  let newKey = generateRandomString();
  urlDatabase[newKey] = address;
  // console.log(urlDatabase);
  return newKey;
};

app.listen(PORT, () => {
  console.log(`Example app listening on port ${PORT}!`);
});

//
// app.get("/fetch", (req, res) => {
//   res.send(`a = ${a}`);
// });
//
/*
* GET codes
*/
app.get("/urls", (req, res) => {
  let template = { urls: urlDatabase, username: req.cookies["username"]};
  
  res.render("urls_index", template);
  //ejs knows to automatically look for ejs in the views folder
});

app.get("/urls/new", (req, res) => {
  let templateVar = req.cookies["username"];
  res.render("urls_new", { username: templateVar});
});

app.get("/u/:shortURL", (req, res) => {
  let longURL = urlDatabase[req.params.shortURL];
  res.redirect(longURL);
});

app.get("/urls/:shortURL", (req, res) => {
  let input = req.params.shortURL;
  let template = { shortURL: input, longURL: urlDatabase[input], username: req.cookies["username"]};
  // console.log(input);
  // console.log(`/urls/:shortURL ${req.params.shortURL} \n for ${urlDatabase[input]}`);

  res.render("urls_show", template);
  //ejs knows to automatically look for ejs in the views folder
});

app.get("/hello", (requ, resp) => {
  let template = { greeting : 'Hello World'};
  resp.render("hello_world", template);
});

//send respond about getting url for selected routes
//
app.get("/set", (req, res) => {
  const a = 1;
  res.send(`a = ${a}`);
});
//

app.get("/", (req, res) => {
  res.redirect('/urls');
  // res.render("hello_world", { greeting: "Hello World"});
});

// return the JSON file of the URL database
app.get("/urls.json", (req, res) => {
  res.json(urlDatabase);
});

app.post("/urls/:id", (req, res) => {
  // console.log(urlDatabase, req.params.id);

  if (urlDatabase === {}) {
    res.end("There is no url in the database");
  }

  if (req.params.id in urlDatabase) {
    urlDatabase[req.params.id] = req.body["newLongURL"];
  }

  // console.log(urlDatabase);
  res.redirect('/urls'); //s/b 302?
});

//Function to add a new URL to the database with a randomized ID
app.post("/urls", (req, res) => {
  // console.log(req.body);  // Log the POST request body to the console
  let tempURL = 'urls/' + addURL(req.body["longURL"]);
  console.log(`new URL: ${tempURL}`);  // Log the POST request body to the console

  res.redirect(tempURL);
  // res.send("Ok");         // Respond with 'Ok' (we will replace this)
});

app.post("/urls/:shortURL/delete", (req, res) => {
  console.log(urlDatabase, req.params.shortURL);

  if (urlDatabase === {}) {
    res.end("There is no url in the database");
  }

  if (req.params.shortURL in urlDatabase) {
    delete urlDatabase[req.params.shortURL];
  }

  res.redirect('/urls'); //s/b 302?
});

app.post("/login", (req, res) => {
  // console.log(req.body["username"]);  // Log the POST request body to the console
 
  res.cookie('username', req.body["username"]);
  res.redirect(302, '/urls');
});

app.post("/logout", (req, res) => {
  // console.log(req.body["username"]);  // Log the POST request body to the console
 
  res.clearCookie("username");
  res.redirect(302, '/urls');
});
 
app.post("*", (req, res) => {
  res.redirect(302, '/urls');
});

express_server.js
const bcrypt = require('bcrypt');
const express = require('express');
const cookieParser = require("cookie-parser");
const cookieSession = require('cookie-session');
const bodyParser = require('body-parser');
const {
  getUserByEmail,
  generateRandomString,
  regOK,
  urlsForUser,
  urlsDetForUser,
  trackVisitors,
  userIDExist } = require('./helpers');
const methodOverride = require('method-override');
const app = express();
const PORT = 8080;
//setup

const {
  users,
  urlDatabase,
  visitorDatabase } = require('./constant');
// ACTION: comment out if pre-set data is not required

app.set("view engine", "ejs");
//setup Express app to use ejs as templating engine

app.use(bodyParser.urlencoded({extended: true}));
//setup body parser

app.use(cookieParser());
//setup cookie parser for visitor cookies

app.use(cookieSession({
  name: 'users',
  keys: ['12345'],
  maxAge: 24 * 60 * 60 * 1000,
  expires: true
}));
//setup the cookie for name, key and 24 hours maximum session stay if session is still open

const expiration = new Date();

app.use(cookieSession({
  name: 'visitors',
  keys: ['67890'],
  maxAge: 30 * 60 * 1000,
  expires: expiration.setTime(Date.now() + 2 * 60 * 1000)
}));
//setup the cookie for name, key and 24 hours maximum session stay if session is still open

app.use(methodOverride('_method'));
//setup method override


// DELETE route codes
//allow user to delete
app.delete("/urls/:shortURL", (req, res) => {
  if (userIDExist(req.session.userID, users) && urlDatabase.length !== 0) {
    if (req.params.shortURL in urlsForUser(req.session.userID, urlDatabase)) {
      delete urlDatabase[req.params.shortURL];
      res.redirect('/urls');
    } else {
      res.redirect(401, 'Not authorized to access /urls');
      //specific case where use does not own the URL
    }
    //end of checking authorized user
  } else {
    res.redirect(401, '/login');
  }
});

// GET routes codes
// Create new url for authorized users
app.get("/urls/new", (req, res) => {
  (userIDExist(req.session.userID, users)) ?
    res.render("urls_new", { email: userIDExist(req.session.userID, users) }) :
    res.redirect('/login');
});

//any user can visit a shorterned URL provided by an authorized user
app.get("/u/:shortURL", (req, res) => {
  if (req.params.shortURL in urlDatabase) {
    // if (req.session.user_id) {
    //   trackVisitors(visitorDatabase, req.params.shortURL, req.session.user_id);
    // } else {
      req.session.user_id = generateRandomString(5);
      console.log(req.session.user_id);
      console.log(req.body);
    //   trackVisitors(visitorDatabase, req.params.shortURL, req.visitor.user_id);
    // }

    res.redirect(urlDatabase[req.params.shortURL].longURL);
  } else {
    res.redirect(404,'/urls');
  }
});

app.get("/urls/:shortURL", (req, res) => {
  if (userIDExist(req.session.userID, users)) {
    if (req.params.shortURL in urlsForUser(req.session.userID, urlDatabase)) {
      let templateVar = {
        shortURL: req.params.shortURL,
        longURL: urlsForUser(req.session.userID, urlDatabase)[req.params.shortURL],
        email: userIDExist(req.session.userID, users)
      };
      res.render("urls_show", templateVar);
      //url exists and user authorized to view
    } else if (req.params.shortURL in urlDatabase) {
      res.redirect(401, '/urls');
      //URL not in authorized list of shortened URL of user
    } else {
      res.redirect(404, '/urls');
      //URL does not exist
    }
    //end of checking authorized users
  } else {
    res.redirect(401, '/login');
  }
});

// return the JSON file of the URL database
app.get("/urls.json", (req, res) => (userIDExist(req.session.userID, users)) ? res.json(urlDatabase) : res.redirect('/login'));

// redirect to index page with the url table
app.get("/urls", (req, res) => {
  console.log(userIDExist(req.session.userID, users));
  if (userIDExist(req.session.userID, users)) {
    let templateVar = {
      urls: urlsDetForUser(req.session.userID, urlDatabase),
      email: userIDExist(req.session.userID, users)
    };
    res.render("urls_index", templateVar);
  //ejs knows to automatically look for ejs in the views folder
  } else {
    res.redirect(401, '/login');
  }
});

// send user to the register page upon clicking the button
app.get("/register", (req, res) => {
  (req.session.userID) ?
    res.redirect('/urls') :
    res.render("urls_login", { status: 200, email: undefined });
});

// send user to the register page upon clicking the button if not logged in
app.get("/login", (req, res) => {
  (req.session.userID) ?
    res.redirect('/urls') :
    res.render("urls_login", { status: 400, email: undefined });
});

//root directory to redirect based on authorization
//default is index (/urls) versus login for unauthorized users
app.get("/", (req, res) => {
  (req.session.userID) ?
    res.redirect('/urls') :
    res.redirect('login');
});

//default to catch all other GETs routes
app.get("*", (req, res) => {
  if (userIDExist(req.session.userID, users)) {
    res.redirect('/urls');
  } else {
    res.redirect('/login');
  }
});

// POST routes codes
//Function to add a new URL to the database with a randomized ID
//then redirect the edit page
app.post("/urls", (req, res) => {
  console.log(userIDExist(req.session.userID, users));
  if (userIDExist(req.session.userID, users)) {
    let newKey = generateRandomString(7);
    urlDatabase[newKey] = {};
    urlDatabase[newKey].longURL = req.body["longURL"];
    urlDatabase[newKey].userID = req.session.userID;
    urlDatabase[newKey].created = Date.now().toDateString();

    res.redirect('urls/' + newKey);
  } else {
    res.redirect(401, '/login');
  }
});

// user login
app.post("/login", (req, res) => {
  let sysId = getUserByEmail(req.body["address"], users);


  if (Object.keys(sysId).length === 0 || req.body["password"] === '') {
    res.redirect(401, 'incorrect input /login');
    //login in error due to a) new user or empty password
  } else if (!bcrypt.compareSync(req.body["password"], sysId["password"])) {
    res.redirect(401, 'incorrect password /login');
    //login in error due to incorrect password by existing user
  } else {
    req.session.userID = sysId["id"];  console.log(sysId);
    res.redirect('/urls');
    //successful login
  }
});

//user logout
app.post("/logout", (req, res) => {
  req.session.userID = null;
  res.redirect(202, '/urls');
});

//user registration
app.post("/register", (req, res) => {
  if (regOK(req.body["address"], req.body["password"]) === false) {
    res.redirect(401, '/register');
    //incorrect input error as backdrop for http form failure
  } else if (Object.keys(getUserByEmail(req.body["address"], users)).length !== 0) {
    res.redirect(401, '/login');
    //user email exists and redirect to login
  } else {
    //user is new with correct inputs and being added with a hashed password
    let newUser = {
      id: generateRandomString(5),
      email: req.body["address"],
      password: bcrypt.hashSync(req.body["password"], 10)
    // hashing with salt for 10 rounds
    };
    users[newUser["id"]] = newUser;
    req.session.userID = newUser.id;
    res.redirect('/urls');
  }
});
 
//default to catch all other POST routes
app.post("*", (req, res) => {
  if (userIDExist(req.session.userID, users)) {
    res.redirect('/urls');
  } else {
    res.redirect('/login');
  }
});

// PUT route codes
// allow (authorized) user to edit an existing long URL
app.put("/urls/:id", (req, res) => {
  if (userIDExist(req.session.userID, users)) {
    // console.log(req.params);
    if (req.params.id in urlsForUser(req.session.userID, urlDatabase)) {
      urlDatabase[req.params.id]["longURL"] = req.body["newLongURL"];
      res.redirect('/urls');
    } else {
      res.redirect(401, 'Not authorized to access /urls');
      //specific case where use does not own the URL
    }
    //end of checking authorized users
  } else {
    res.redirect(401, '/login');
  }
});

//listener
app.listen(PORT, () => {
  console.log(`Example app listening on port ${PORT}!`);
});